parser/parser.go

```diff
@@ -0,0 +1,393 @@
+package parser
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/alecthomas/participle"
+	"github.com/alecthomas/participle/lexer"
+	"github.com/flynn/biscuit-go"
+	"github.com/flynn/biscuit-go/datalog"
+)
+
+var (
+	ErrVariableInFact = errors.New("parser: a fact cannot contain any variable")
+)
+
+var DefaultParserOptions = []participle.Option{
+	participle.Lexer(lexer.DefaultDefinition),
+	participle.UseLookahead(3),
+}
+
+type Parser interface {
+	Fact() FactParser
+	Rule() RuleParser
+	Caveat() CaveatParser
+}
+
+type FactParser interface {
+	MustParse(fact string) biscuit.Fact
+	Parse(fact string) (biscuit.Fact, error)
+}
+
+type RuleParser interface {
+	MustParse(rule string) biscuit.Rule
+	Parse(rule string) (biscuit.Rule, error)
+}
+
+type CaveatParser interface {
+	MustParse(caveat string) biscuit.Caveat
+	Parse(caveat string) (biscuit.Caveat, error)
```

Comment:
What do you think about exporting these all as top level package functions? (`ParseFact`, `MustParseFact`, etc.)

Reply:
Actually, maybe we should do something like:

```go
type Parser interface {
  Fact(fact string) (biscuit.Fact, error)
  Rule(rule string) (biscuit.Rule, error)
  Caveat(caveat string) (biscuit.Caveat, error)
  Must() MustParser
}

type MustParser interface{
  Fact(fact string) biscuit.Fact
  Rule(rule string) biscuit.Rule
  Caveat(caveat string) biscuit.Caveat
}
```

Reply:
Ah thanks! My initial version looked like this, but couldn't figure how to add the "Must" version properly. I like this one. I'd vote to keep the struct / interface rather than package functions to ease testing in some context. Is there any cons ?

Reply:
I don't see a downside, let's stick with interfaces.